internal/storage/metastorage/firestore/event_storage.go

```diff
@@ -42,47 +53,413 @@ func newEventStorage(params Params, client *firestore.Client) (internal.EventSto
+			if err != nil {
+				return nil, xerrors.Errorf("failed to parse event entry from firestore document: %w", err)
+			}
+		}
+		return eventEntries, nil
+	})
+}
+
+func (e *eventStorageImpl) getEventsByBlockHeight(ctx context.Context, eventTag uint32, blockHeight uint64) ([]*model.EventEntry, error) {
+	docs, err := e.client.
+		Collection("env").Doc(e.env).
+		Collection("events").Query.
+		Where("EventTag", "==", eventTag).
+		Where("BlockHeight", "==", int64(blockHeight)).
+		Documents(ctx).GetAll()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get event entries: %w", err)
+	}
+	eventEntries := make([]*model.EventEntry, len(docs))
+	for i, doc := range docs {
+		eventEntries[i], err = e.intoEventEntry(doc)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to parse event entry from firestore document: %w", err)
+		}
+	}
+	return eventEntries, nil
 }
 
 // GetEventsByBlockHeight implements internal.EventStorage.
-func (*eventStorageImpl) GetEventsByBlockHeight(ctx context.Context, eventTag uint32, blockHeight uint64) ([]*model.EventEntry, error) {
-	panic("unimplemented")
+func (e *eventStorageImpl) GetEventsByBlockHeight(ctx context.Context, eventTag uint32, blockHeight uint64) ([]*model.EventEntry, error) {
+	if err := e.validateEventTag(eventTag); err != nil {
+		return nil, err
+	}
+	return e.instrumentGetEventsByBlockHeight.Instrument(ctx, func(ctx context.Context) ([]*model.EventEntry, error) {
+		items, err := e.getEventsByBlockHeight(ctx, eventTag, blockHeight)
+		if len(items) == 0 {
+			return nil, errors.ErrItemNotFound
+		}
+		return items, err
+	})
 }
 
 // GetEventsByEventIdRange implements internal.EventStorage.
-func (*eventStorageImpl) GetEventsByEventIdRange(ctx context.Context, eventTag uint32, minEventId int64, maxEventId int64) ([]*model.EventEntry, error) {
-	panic("unimplemented")
+func (e *eventStorageImpl) GetEventsByEventIdRange(ctx context.Context, eventTag uint32, minEventId int64, maxEventId int64) ([]*model.EventEntry, error) {
+	if err := e.validateEventTag(eventTag); err != nil {
+		return nil, err
+	}
+	return e.instrumentGetEventsByEventIdRange.Instrument(ctx, func(ctx context.Context) ([]*model.EventEntry, error) {
+		docs, err := e.client.
+			Collection("env").Doc(e.env).
+			Collection("events").Query.
+			StartAt(e.getEventDocRef(eventTag, minEventId)).
+			EndBefore(e.getEventDocRef(eventTag, maxEventId)).
+			OrderBy(firestore.DocumentID, firestore.Asc).
+			Documents(ctx).GetAll()
+		if err != nil {
+			return nil, xerrors.Errorf("failed to get event entries: %w", err)
+		}
+		if len(docs) != int(maxEventId-minEventId) {
+			return nil, errors.ErrItemNotFound
+		}
+		eventEntries := make([]*model.EventEntry, len(docs))
+		for i, doc := range docs {
+			eventEntries[i], err = e.intoEventEntry(doc)
+			if err != nil {
+				return nil, xerrors.Errorf("failed to parse event entry from firestore document: %w", err)
+			}
+			if eventEntries[i].EventId != minEventId+int64(i) {
```

Comment:
why do we need to calculate the EventId, don't we already have it in the docs?

Reply:
This is different from dynamodb, in firestore we're using range query instead of batchGet. So we check the continuity of the docs, if it's not then some event is missing.

Reply:
if we use batchGet, we also get the continuity check, since a missing event will result in a not found. right?